Add controller framework

shague opened the pull request:
This adds the initial controller framework as well as Makefile and Travis support. The etcd interactions will be added in subsequent patches.

shague-art committed:
add IDE files to .gitignore

Signed-off-by: Sam Hague <shague@gmail.com>

shague-art committed:
Ignore bin build directory

Signed-off-by: Sam Hague <shague@gmail.com>

shague-art committed:
Add Makefile targets

Signed-off-by: Sam Hague <shague@gmail.com>

shague-art committed:
Add controller framework

Signed-off-by: Sam Hague <shague@gmail.com>

shague-art committed:
Add travis support

Signed-off-by: Sam Hague <shague@gmail.com>

trozet commented on cmd/stargazer/main.go:
```diff
@@ -0,0 +1,117 @@
+		os.Exit(0)
+	}
+
+	//cfg := new(config.Config)
```
remove this line

trozet commented on cmd/stargazer/main.go:
```diff
@@ -0,0 +1,117 @@
+	// TODO: ensure connection to etcd is available
+
+	controllerCtrl := &controllerControl{
+		ctx:              ctx,
```
one thing is the google docs say not to store context inside of structs. I did the same thing in my code, not sure if we should change it:

https://github.com/golang/go/issues/22602
https://groups.google.com/forum/#!topic/golang-nuts/xRbzq8yzKWI

shague replied:
Those comments sort of make sense, but may not apply here. There is only one function for controller, NewNodeController and the ctx is passed in via the function call as the comments want. Then the thread is started so there are no other function calls to pass a context. I can't see this API changing much so I think we are good here. For other cases, I agree we should determine what are pattern should be.

trozet replied:
Let's leave it until later then.

trozet commented on cmd/stargazer/main.go:
```diff
@@ -0,0 +1,117 @@
+	ctx              context.Context
+	controllerStates map[string]*controllerState
+	config           *config.Config
+	stop             chan struct{}
```
In what cases do you think you will need to use this?

shague replied:
It is used to sync with the cache and informer pieces. in a later patch when the etcd is added, there will also be a cache and informer piece. The controller will need to wait till the cache is ready before proceeding.

trozet commented on pkg/controllers/node/node_controller.go:
```diff
@@ -0,0 +1,57 @@
+	"sync"
+)
+
+type NodeController struct {
```
should this be composed from a generic Controller struct? That way when you add other controllers they can all just inherit from that type?

shague replied:
Some of it might be common, like context and config. There does need to be a distinct structure though since each controller is different an has some unique data. When I add more meat to the node controller and add the unp controller I will look at refactoring common parts.

trozet commented on cmd/stargazer/main.go:
```diff
@@ -0,0 +1,117 @@
+}
+
+// Object for keeping track of Controller information.
+type controllerState struct {
```
This name seems like it doesn't really fit. This doesn't maintain a state variable or anything, it just tracks a controller and how many threads it supports?

shague replied:
The Controller is just an interface with a Run function. This is the controller itself, so it sounds like that is state since it's all I have. I can't think of a better name but am not tied to it either.

trozet replied:
Any struct which implements Run will be a Controller. You already do that with Node Controller. This is just a struct holding a controller and threadiness. If your goal is for this object to be a controller, then I think it would be better to just include your threadiness field inside of the controller struct itself, and then you don't need this struct at all.

shague replied:
Renamed ControllerState to ControllerInfo since it holds the controller along with the request number of workers. This ControllerInfo is used with the ControllerControl map of controllerType,Info to start each controller type requested via the configuration.

trozet commented on pkg/controllers/node/node_controller.go:
```diff
@@ -0,0 +1,57 @@
+	nodemapLock  sync.Mutex
+}
+
+func NewNodeController(ctx context.Context, cfg *config.Config) controller.Controller {
```
may want to comment on all the functions so we generate proper go docs later

shague replied:
OK, will add more comments. I admit I don't see much value in a commend like NewNodeController creates a new Node Controller since that is the same as the function name.

trozet reviewed (changes requested):
Minor comments, i think its fine to approve after that

trozet commented:
Looks like the travis stuff is passing even though their are lint failures? maybe not exiting with non-zero return code
https://travis-ci.com/nimbess/stargazer/builds/117186434#L477

trozet commented on Dockerfile:
```diff
@@ -0,0 +1,6 @@
+FROM golang:alpine
```
should we use golang:1.11 image here? Would match the golang version we are targeting.

trozet replied:
oh wait you compile it then add it into this container. Should you compile it in another container like golang:1.11 i mean.

shague replied:
Yes I plan to add another target to compile it in a container. This target here is really for the IT testing - this target here is just a golang container that we launch the binary into to see if it works.

shague-art committed:
Add updates for review comments

Signed-off-by: Sam Hague <shague@gmail.com>

shague-art committed:
Update tests after review comments

Signed-off-by: Sam Hague <shague@gmail.com>

shague commented:
Also updated the .travis.yml to include the set -e so it fails. Not needed this patchset, though, as the checks passed with no issues because I fixed the previous issues found.
